src/v/raft/recovery_stm.cc

```diff
@@ -557,6 +557,18 @@ ss::future<> recovery_stm::replicate(
         _stop_requested = true;
         return ss::now();
     }
+    if (meta.value()->expected_log_end_offset >= _last_batch_offset) {
+        vlog(
+          _ctxlog.trace,
+          "follower expected log end offset is already updated, stopping "
+          "recovery. Expected log end offset: {}, recovery range last offset: "
+          "{}",
+          meta.value()->expected_log_end_offset,
+          _last_batch_offset);
+
+        _stop_requested = true;
+        return ss::now();
+    }
```

Comment:
As far as I can tell, `_last_batch_offset` is only updated in `read_range_for_recovery`, and `expected_log_end_offset` is updated after this new statement. So when can this happen? I don't understand the scenario, as this would seem that no batches were read or we are reading duplicate batches (which is maybe the case from how `next_index` is incremented that overlapping ranges can be read?). Can you help explain on when we'd get batches here when no new data is read?

Reply:
Also this doesn't have a chance to stop recovery early right? I don't see how this can guard against that. The bytes limit in `read_range_for_recovery` could limit where no new data is read but there is still new data in the log that needs to be sent for recovery based on how the start offset is incremented.

Reply:
The `expected_log_end_offset` is a field representing last offset of the follower log after it accepts all pending `append_entries` requests. This piece of the state is used mainly to coordinate the `replicate_entries_stm` with `recovery_stm`. i.e. if a message was already sent (or is about to be sent) from `replicate_entries_stm` we should skip sending it from `recovery_stm`.

Reply:
Ack, thanks for the explaination I see it now 😄